support-frontend/app/controllers/RedemptionController.scala

```diff
@@ -0,0 +1,181 @@
+package controllers
+
+import actions.CustomActionBuilders
+import actions.CustomActionBuilders.AuthRequest
+import admin.settings.{AllSettings, AllSettingsProvider}
+import assets.{AssetsResolver, RefPath, StyleContent}
+import cats.data.EitherT
+import cats.implicits._
+import com.gu.identity.model.{User => IdUser}
+import com.gu.support.redemptions.CorporateCustomer
+import com.gu.support.redemptions.redemptions.RedemptionCode
+import controllers.RedemptionController._
+import io.circe.syntax._
+import play.api.libs.circe.Circe
+import play.api.mvc.{AbstractController, Action, AnyContent, ControllerComponents, Result}
+import services.{IdentityService, TestUserService}
+import views.EmptyDiv
+import views.html.helper.CSRF
+import views.html.subscriptionRedemptionForm
+import cats.implicits._
+import com.gu.monitoring.SafeLogger
+import SafeLogger._
+import lib.RedirectWithEncodedQueryString
+import play.twirl.api.Html
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class RedemptionController(
+  val actionRefiners: CustomActionBuilders,
+  val assets: AssetsResolver,
+  settingsProvider: AllSettingsProvider,
+  identityService: IdentityService,
+  testUsers: TestUserService,
+  components: ControllerComponents,
+  fontLoaderBundle: Either[RefPath, StyleContent]
+)(
+  implicit val ec: ExecutionContext
+) extends AbstractController(components) with Circe {
+
+  import actionRefiners._
+
+  implicit val a: AssetsResolver = assets
+
+  implicit val settings: AllSettings = settingsProvider.getAllSettings()
+  val title = "Support the Guardian | Redeem your code"
+  val id = EmptyDiv("subscriptions-redemption-page")
+  val js = "subscriptionsRedemptionPage.js"
+  val css = "digitalSubscriptionCheckoutPage.css" //TODO: Don't need this?
+
+  def displayForm(redemptionCode: String): Action[AnyContent] = NoCacheAction().async {
+    implicit request =>
+      getCorporateCustomer(redemptionCode).value.map(
+        customerOrError =>
+          Ok(subscriptionRedemptionForm(
+            title = title,
+            mainElement = id,
+            js = js,
+            css = css,
+            fontLoaderBundle = fontLoaderBundle,
+            csrf = None,
+            uatMode = false,
+            stage = "checkout",
+            redemptionCode = redemptionCode,
+            customerOrError = customerOrError,
+            user = None,
+            submitted = false
+          ))
+      )
+
+  }
+
+  def displayThankYou(stage: String): Action[AnyContent] = authenticatedAction(subscriptionsClientId).async {
+    implicit request: AuthRequest[Any] =>
+      identityService.getUser(request.user.minimalUser).fold(
+        error => thankYouPage(stage, None),
```

Comment:
re changed the way this works, looks good
```suggestion
        error => {
          SafeLogger.error(scrub"failed to show user on thank you page. Error was - ${error}")
          thankYouPage(stage, None)
        },
```
don't thint it's critical enough to send it to the urgent 9-5 cloudwatch alarm, but make your own mind up based on what the effect woudl be.